// @elizaos/plugin-hyperliquid/dist/index.js
// src/actions/spotTrade.ts
import {
  logger,
  ModelType,
  parseKeyValueXml
} from "@elizaos/core";
import { Hyperliquid } from "hyperliquid";

// src/types.ts
import { z } from "zod";
var SpotOrderSchema = z.object({
  coin: z.string().min(1),
  is_buy: z.boolean(),
  sz: z.number().positive(),
  limit_px: z.number().positive().nullable(),
  reduce_only: z.boolean().default(false),
  order_type: z.object({
    limit: z.object({
      tif: z.enum(["Ioc", "Gtc"])
    })
  }).default({ limit: { tif: "Gtc" } })
});
var HyperliquidError = class extends Error {
  constructor(message, code, details) {
    super(message);
    this.code = code;
    this.details = details;
    this.name = "HyperliquidError";
  }
};
var PRICE_VALIDATION = {
  MARKET_ORDER: {
    MIN_RATIO: 0.5,
    // -50% from mid price
    MAX_RATIO: 1.5
    // +50% from mid price
  },
  LIMIT_ORDER: {
    WARNING_MIN_RATIO: 0.2,
    // -80% from mid price
    WARNING_MAX_RATIO: 5
    // +500% from mid price
  },
  SLIPPAGE: 0.01
  // 1% slippage for market orders
};

// src/templates.ts
var spotTradeTemplate = `<task>
Look at your LAST RESPONSE in the conversation where you confirmed a trade request.
Based on ONLY that last message, extract the trading details.
</task>

<context>
Recent conversation:
{{recentMessages}}
</context>

<rules>
For Hyperliquid spot trading:

Market orders (executes immediately at best available price):
- "buy 1 HYPE" -> coin: HYPE, is_buy: true, sz: 1
- "sell 2 HYPE" -> coin: HYPE, is_buy: false, sz: 2
- "market buy 1 HYPE" -> coin: HYPE, is_buy: true, sz: 1
- "market sell 2 HYPE" -> coin: HYPE, is_buy: false, sz: 2

Limit orders (waits for specified price):
- "buy 1 HYPE at 20 USDC" -> coin: HYPE, is_buy: true, sz: 1, limit_px: 20
- "sell 0.5 HYPE at 21 USDC" -> coin: HYPE, is_buy: false, sz: 0.5, limit_px: 21
- "limit buy 1 HYPE at 20 USDC" -> coin: HYPE, is_buy: true, sz: 1, limit_px: 20
- "limit sell 0.5 HYPE at 21 USDC" -> coin: HYPE, is_buy: false, sz: 0.5, limit_px: 21
</rules>

<instructions>
1. Extract the coin symbol (HYPE, ETH, etc.) - just the symbol, no suffixes
2. Determine if it's a buy (true) or sell (false) order
3. Extract the size/quantity (sz) to trade as specified in the message
4. Extract limit_px only if a price was specified with "at X USDC", otherwise set to null for market orders
5. Words like "market" or "limit" at the start are optional but help clarify intent
</instructions>

Respond with an XML block containing only the extracted values. Use key-value pairs.

Example response for a market buy:
<response>
    <coin>HYPE</coin>
    <is_buy>true</is_buy>
    <sz>1</sz>
    <limit_px>null</limit_px>
</response>

Example response for a limit sell:
<response>
    <coin>HYPE</coin>
    <is_buy>false</is_buy>
    <sz>0.5</sz>
    <limit_px>21</limit_px>
</response>`;
var priceCheckTemplate = `<task>
Look at your LAST RESPONSE in the conversation where you confirmed which token price to check.
Based on ONLY that last message, extract the token symbol.
</task>

<context>
Recent conversation:
{{recentMessages}}
</context>

<examples>
- "I'll check PIP price for you" -> symbol: PIP
- "Let me check the price of HYPE" -> symbol: HYPE
- "I'll get the current ETH price" -> symbol: ETH
</examples>

<instructions>
1. Extract the token symbol from your last message
2. Return just the token symbol (PIP, HYPE, ETH, etc.)
3. Remove any suffixes like "-SPOT" or "USDC"
4. If multiple tokens are mentioned, use the last one
</instructions>

Respond with an XML block containing only the extracted value.

Example response:
<response>
    <symbol>HYPE</symbol>
</response>`;

// src/actions/spotTrade.ts
var spotTrade = {
  name: "SPOT_TRADE",
  similes: ["SPOT_ORDER", "SPOT_BUY", "SPOT_SELL"],
  description: "Place a spot trade order on Hyperliquid",
  validate: async (runtime) => {
    return !!runtime.getSetting("HYPERLIQUID_PRIVATE_KEY");
  },
  handler: async (runtime, message, state, _options, callback) => {
    try {
      const currentState = state || await runtime.composeState(message);
      const contextPrompt = spotTradeTemplate.replace(
        "{{recentMessages}}",
        currentState.recentMessagesString || ""
      );
      const modelResult = await runtime.useModel(ModelType.TEXT_SMALL, {
        prompt: contextPrompt
      });
      const parsedContent = parseKeyValueXml(modelResult);
      if (!parsedContent || Object.keys(parsedContent).length === 0) {
        throw new HyperliquidError(
          "Could not parse trading parameters from conversation"
        );
      }
      const orderParams = {
        coin: parsedContent.coin,
        is_buy: parsedContent.is_buy === "true",
        sz: parseFloat(parsedContent.sz),
        limit_px: parsedContent.limit_px === "null" ? null : parseFloat(parsedContent.limit_px)
      };
      logger.info(
        "Parsed order parameters:",
        JSON.stringify(orderParams, null, 2)
      );
      const validatedOrder = SpotOrderSchema.parse(orderParams);
      logger.info("Validated order:", validatedOrder);
      const sdk = new Hyperliquid({
        privateKey: runtime.getSetting("HYPERLIQUID_PRIVATE_KEY"),
        testnet: runtime.getSetting("HYPERLIQUID_TESTNET") === "true",
        enableWs: false
      });
      await sdk.connect();
      const [meta, assetCtxs] = await sdk.info.spot.getSpotMetaAndAssetCtxs();
      const tokenIndex = meta.tokens.findIndex(
        (token) => token.name.toUpperCase() === validatedOrder.coin.toUpperCase()
      );
      if (tokenIndex === -1) {
        throw new HyperliquidError(
          `Could not find token ${validatedOrder.coin}`
        );
      }
      const tokenInfo = meta.tokens[tokenIndex];
      logger.info("Found token:", tokenInfo.name);
      const marketIndex = assetCtxs.findIndex(
        (ctx) => ctx.coin === `${validatedOrder.coin}-SPOT`
      );
      if (marketIndex === -1) {
        throw new HyperliquidError(
          `Could not find market for ${validatedOrder.coin}`
        );
      }
      const marketCtx = assetCtxs[marketIndex];
      if (!marketCtx || !marketCtx.midPx) {
        throw new HyperliquidError(
          `Could not get market price for ${validatedOrder.coin}`
        );
      }
      const midPrice = Number(marketCtx.midPx);
      const isMarketOrder = !validatedOrder.limit_px;
      let finalPrice;
      if (isMarketOrder) {
        const slippage = PRICE_VALIDATION.SLIPPAGE;
        finalPrice = validatedOrder.is_buy ? midPrice * (1 + slippage) : midPrice * (1 - slippage);
        if (finalPrice < midPrice * PRICE_VALIDATION.MARKET_ORDER.MIN_RATIO || finalPrice > midPrice * PRICE_VALIDATION.MARKET_ORDER.MAX_RATIO) {
          throw new HyperliquidError(
            `Market order price (${finalPrice.toFixed(2)} USDC) is too far from market price (${midPrice.toFixed(2)} USDC). This might be due to low liquidity.`
          );
        }
      } else {
        finalPrice = validatedOrder.limit_px;
        if (validatedOrder.is_buy && finalPrice > midPrice) {
          throw new HyperliquidError(
            `Cannot place buy limit order at ${finalPrice.toFixed(2)} USDC because it's above market price (${midPrice.toFixed(2)} USDC). To execute immediately, use a market order. For a limit order, set a price below ${midPrice.toFixed(2)} USDC.`
          );
        } else if (!validatedOrder.is_buy && finalPrice < midPrice) {
          throw new HyperliquidError(
            `Cannot place sell limit order at ${finalPrice.toFixed(2)} USDC because it's below market price (${midPrice.toFixed(2)} USDC). To execute immediately, use a market order. For a limit order, set a price above ${midPrice.toFixed(2)} USDC.`
          );
        }
        if (finalPrice < midPrice * PRICE_VALIDATION.LIMIT_ORDER.WARNING_MIN_RATIO || finalPrice > midPrice * PRICE_VALIDATION.LIMIT_ORDER.WARNING_MAX_RATIO) {
          logger.warn(
            `Limit price (${finalPrice.toFixed(2)} USDC) is very different from market price (${midPrice.toFixed(2)} USDC). Make sure this is intentional.`,
            {
              finalPrice,
              midPrice,
              ratio: finalPrice / midPrice
            }
          );
        }
      }
      const rounded_px = Number(finalPrice.toFixed(tokenInfo.szDecimals));
      const orderRequest = {
        coin: `${validatedOrder.coin}-SPOT`,
        asset: 1e4 + marketIndex,
        is_buy: validatedOrder.is_buy,
        sz: validatedOrder.sz,
        limit_px: rounded_px,
        reduce_only: false,
        order_type: isMarketOrder ? { market: {} } : { limit: { tif: "Gtc" } }
      };
      logger.info("Placing order:", orderRequest);
      const result = await sdk.exchange.placeOrder(orderRequest);
      if (result.status === "ok" && result.response?.type === "order" && result.response.data?.statuses?.[0]?.error) {
        throw new HyperliquidError(result.response.data.statuses[0].error);
      }
      if (callback) {
        const action = validatedOrder.is_buy ? "buy" : "sell";
        const executionPrice = result.response?.data?.statuses?.[0]?.px || rounded_px;
        callback({
          text: `Successfully placed ${isMarketOrder ? "a market" : "a limit"} order to ${action} ${validatedOrder.sz} ${validatedOrder.coin} at ${executionPrice}`,
          content: result
        });
      }
      return true;
    } catch (error) {
      logger.error("Error placing spot order:", error);
      const errorMessage = error instanceof Error ? error.message : String(error);
      if (callback) {
        callback({
          text: `Error placing spot order: ${errorMessage}`,
          content: { error: errorMessage }
        });
      }
      return false;
    }
  },
  examples: [
    [
      {
        name: "{{user1}}",
        content: {
          text: "Buy 0.1 HYPE at 20 USDC"
        }
      },
      {
        name: "{{agent}}",
        content: {
          text: "I'll place a spot buy order for 0.1 HYPE at 20 USDC.",
          action: "SPOT_TRADE"
        }
      },
      {
        name: "{{agent}}",
        content: {
          text: "Successfully placed a limit order to buy 0.1 HYPE at 20 USDC"
        }
      }
    ],
    [
      {
        name: "{{user1}}",
        content: {
          text: "Sell 2 HYPE at 21 USDC"
        }
      },
      {
        name: "{{agent}}",
        content: {
          text: "I'll place a spot sell order for 2 HYPE at 21 USDC.",
          action: "SPOT_TRADE"
        }
      },
      {
        name: "{{agent}}",
        content: {
          text: "Successfully placed a limit order to sell 2 HYPE at 21 USDC"
        }
      }
    ]
  ]
};

// src/actions/priceCheck.ts
import {
  logger as logger2,
  ModelType as ModelType2,
  parseKeyValueXml as parseKeyValueXml2
} from "@elizaos/core";
import { Hyperliquid as Hyperliquid2 } from "hyperliquid";
var priceCheck = {
  name: "PRICE_CHECK",
  similes: ["CHECK_PRICE", "GET_PRICE", "PRICE", "CURRENT_PRICE"],
  description: "Get current price for a token on Hyperliquid",
  validate: async () => true,
  // Public endpoint
  handler: async (runtime, message, state, _options, callback) => {
    try {
      const currentState = state || await runtime.composeState(message);
      const contextPrompt = priceCheckTemplate.replace(
        "{{recentMessages}}",
        currentState.recentMessagesString || ""
      );
      const result = await runtime.useModel(ModelType2.TEXT_SMALL, {
        prompt: contextPrompt
      });
      const parsedContent = parseKeyValueXml2(result);
      if (!parsedContent || !parsedContent.symbol) {
        throw new HyperliquidError(
          "Could not determine which token price to check"
        );
      }
      logger2.info("Checking price for token:", parsedContent.symbol);
      const sdk = new Hyperliquid2({
        enableWs: false
      });
      await sdk.connect();
      const [meta, assetCtxs] = await sdk.info.spot.getSpotMetaAndAssetCtxs();
      const tokenIndex = meta.tokens.findIndex(
        (token) => token.name.toUpperCase() === parsedContent.symbol.toUpperCase()
      );
      if (tokenIndex === -1) {
        throw new HyperliquidError(
          `Could not find token ${parsedContent.symbol}`
        );
      }
      const marketIndex = assetCtxs.findIndex(
        (ctx) => ctx.coin === `${parsedContent.symbol}-SPOT`
      );
      if (marketIndex === -1) {
        throw new HyperliquidError(
          `Could not find market for ${parsedContent.symbol}`
        );
      }
      const marketCtx = assetCtxs[marketIndex];
      if (!marketCtx || !marketCtx.midPx) {
        throw new HyperliquidError(
          `Could not get market price for ${parsedContent.symbol}`
        );
      }
      const price = Number(marketCtx.midPx);
      const dayChange = ((price - Number(marketCtx.prevDayPx)) / Number(marketCtx.prevDayPx) * 100).toFixed(2);
      const volume = Number(marketCtx.dayNtlVlm).toFixed(2);
      if (callback) {
        callback({
          text: `${parsedContent.symbol} price: ${price.toFixed(2)} USDC (24h change: ${dayChange}%, volume: ${volume} USDC)`,
          content: {
            symbol: parsedContent.symbol,
            price,
            dayChange,
            volume
          }
        });
      }
      return true;
    } catch (error) {
      logger2.error("Error checking price:", error);
      const errorMessage = error instanceof Error ? error.message : String(error);
      if (callback) {
        callback({
          text: `Error checking price: ${errorMessage}`,
          content: { error: errorMessage }
        });
      }
      return false;
    }
  },
  examples: [
    [
      {
        name: "{{user1}}",
        content: {
          text: "What's the current price of PIP?"
        }
      },
      {
        name: "{{agent}}",
        content: {
          text: "I'll check the current PIP price for you.",
          action: "PRICE_CHECK"
        }
      },
      {
        name: "{{agent}}",
        content: {
          text: "PIP price: 19.73 USDC (24h change: -1.82%, volume: 1053445.75 USDC)"
        }
      }
    ]
  ]
};

// src/actions/cancelOrders.ts
import {
  logger as logger3
} from "@elizaos/core";
import { Hyperliquid as Hyperliquid3 } from "hyperliquid";
var cancelOrders = {
  name: "CANCEL_ORDERS",
  similes: ["CANCEL_ALL_ORDERS", "CANCEL", "CANCEL_ALL"],
  description: "Cancel all open orders on Hyperliquid",
  validate: async (runtime) => {
    return !!runtime.getSetting("HYPERLIQUID_PRIVATE_KEY");
  },
  handler: async (runtime, _message, _state, _options, callback) => {
    try {
      const sdk = new Hyperliquid3({
        privateKey: runtime.getSetting("HYPERLIQUID_PRIVATE_KEY"),
        testnet: runtime.getSetting("HYPERLIQUID_TESTNET") === "true",
        enableWs: false
      });
      await sdk.connect();
      logger3.info("Cancelling all open orders...");
      const result = await sdk.custom.cancelAllOrders();
      logger3.info("Cancel result:", result);
      if (callback) {
        const cancelledCount = result?.response?.data?.statuses?.length || 0;
        callback({
          text: cancelledCount > 0 ? `Successfully cancelled ${cancelledCount} open order${cancelledCount > 1 ? "s" : ""}` : "No open orders to cancel",
          content: result
        });
      }
      return true;
    } catch (error) {
      logger3.error("Error cancelling orders:", error);
      const errorMessage = error instanceof Error ? error.message : String(error);
      if (callback) {
        callback({
          text: `Error cancelling orders: ${errorMessage}`,
          content: { error: errorMessage }
        });
      }
      return false;
    }
  },
  examples: [
    [
      {
        name: "{{user1}}",
        content: {
          text: "Cancel all my orders"
        }
      },
      {
        name: "{{agent}}",
        content: {
          text: "I'll cancel all your open orders.",
          action: "CANCEL_ORDERS"
        }
      },
      {
        name: "{{agent}}",
        content: {
          text: "Successfully cancelled 2 open orders"
        }
      }
    ]
  ]
};

// src/index.ts
var hyperliquidPlugin = {
  name: "hyperliquid",
  description: "Hyperliquid plugin",
  actions: [spotTrade, priceCheck, cancelOrders],
  providers: [],
  evaluators: [],
  services: []
};
var index_default = hyperliquidPlugin;
export {
  index_default as default,
  hyperliquidPlugin
};
